source/grid/grid_out.cc

```diff
@@ -2590,51 +2590,70 @@ void GridOut::write_mesh_per_processor_as_vtu (const Triangulation<dim,spacedim>
+                                 ".proc" + Utilities::int_to_string(i, 4) +
+                                 ".vtu");
+
+          const std::string pvtu_master_filename = (filename_without_extension + ".pvtu");
+          std::ofstream pvtu_master (pvtu_master_filename.c_str());
+
+          DataOut<dim, DoFHandler<dim,spacedim> > data_out;
+          data_out.attach_triangulation (*tr);
+
+          // We need a dummy vector with the names of the data values in the .vtu files
+          // in order that the .pvtu contains reference these values
+          Vector<float> dummy_vector (tr->n_active_cells());
+          data_out.add_data_vector (dummy_vector, "level");
+          data_out.add_data_vector (dummy_vector, "subdomain");
+          data_out.add_data_vector (dummy_vector, "level_subdomain");
+          data_out.add_data_vector (dummy_vector, "proc_writing");
+
+          data_out.build_patches ();
+
+          data_out.write_pvtu_record (pvtu_master, filenames);
+        }
+    }
+  else
     {
-      std::vector<std::string> filenames;
-
-      //.pvtu needs to reference the files without a relative path because it will be written
-      //in the same directory. For this, remove any paths from filename.
-      std::size_t pos = filename_without_extension.find_last_of('/');
-      if (pos == std::string::npos)
-        pos = 0;
-      else
-        pos += 1;
-      for (unsigned int i=0; i<Utilities::MPI::n_mpi_processes(MPI_COMM_WORLD); ++i)
-        filenames.push_back (filename_without_extension.substr(pos) +
-                             ".proc" + Utilities::int_to_string(i, 4) +
-                             ".vtu");
-
-      const std::string pvtu_master_filename = (filename_without_extension + ".pvtu");
-      std::ofstream pvtu_master (pvtu_master_filename.c_str());
-
-      DataOut<dim, DoFHandler<dim,spacedim> > data_out;
-      data_out.attach_triangulation (tria);
-
-      //We need a dummy vector with the names of the data values in the .vtu files
-      //in order that the .pvtu contains reference these values
-      Vector<float> dummy_vector (tria.n_active_cells());
-      data_out.add_data_vector (dummy_vector, "level");
-      data_out.add_data_vector (dummy_vector, "subdomain");
-      data_out.add_data_vector (dummy_vector, "level_subdomain");
-      data_out.add_data_vector (dummy_vector, "proc_writing");
-
-      data_out.build_patches ();
-
-      data_out.write_pvtu_record (pvtu_master, filenames);
+      const std::string new_file = (filename_without_extension + ".vtu");
+      std::ofstream out(new_file.c_str());
+      std::vector<std::tuple<unsigned int, unsigned int, std::string> > vector_data_ranges;
+      DataOutBase::VtkFlags flags;
+      DataOutBase::write_vtu (patches,
+                              data_names,
+                              vector_data_ranges,
+                              flags,
+                              out);
```

Comment:
If I have understood things correctly this block is, up to `new_file`, identical to the top of the previous block. Would you mind adding another `if` statement on top (or use the ternary operator) to set up `new_file` and then put this afterwards to avoid the duplication?